stdlib/Random/docs/src/index.md

```diff
@@ -168,9 +168,9 @@ julia> rand(Die, 3)
 
 julia> a = Vector{Die}(undef, 3); rand!(a)
 3-element Vector{Die}:
+ Die(9)
```

Comment:
This change is due to the unfortunate (but not buggy) behavior @mikmoore pointed out, in conjunction with these doctests not heeding the [warning](https://docs.julialang.org/en/v1/stdlib/Random/#Random-Numbers) not to depend on specific random sequences—a warning that is hard to follow in this case.

This PR also breaks doctests in SparseArrays for the same reason. I'm not sure how to edit SparseArrays in a PR to JuliaLang/julia

Reply:
We can open a PR on `SparseArrays.jl` side to make its doctest independent from the global RNG state by fixing seed manually.
Then it would pass with/without changes here.

Reply:
My suspicion was right. Documenter.jl mutates the random state on this PR by converting REPL output to a string. The results are inconsistent between evaluation in Main and running doctests because the rng mutations of `show` on this PR depend on the number of exported names of parent modules.

Cross-ref some other show-related doctest issues https://github.com/JuliaDocs/Documenter.jl/issues/1631 & slightly less relevant https://github.com/JuliaDocs/Documenter.jl/issues/1580.

While it is technically fair game to mutate random state in `show`, it feels like a lot to ask the folks at Documenter to deal with just so that we can randomize pivot selection.

@mortenpi, do you think it is reasonable for Documenter's doctests to accommodate `show` mutating default rng state? Specifically, showing a type (or anything printed with a type) that is a part of a module that has a lot of exported names will mutate the global random state based on those list(s) of exported names.

In the interim, I have introduced a hack to work around this issue by suppressing rng state mutation in `names`.

<!--
Replacing
```
value === nothing || Base.invokelatest(show, IOContext(buf, :limit => true), MIME"text/plain"(), value)
```
with
```
value === nothing || if VERSION >= v"1.9.0-DEV"
    t = current_task()
    x = t.rngState0, t.rngState1, t.rngState2, t.rngState3
    Base.invokelatest(show, IOContext(buf, :limit => true), MIME"text/plain"(), value)
    t.rngState0, t.rngState1, t.rngState2, t.rngState3 = x
else
    Base.invokelatest(show, IOContext(buf, :limit => true), MIME"text/plain"(), value)
end
```
[here](https://github.com/JuliaDocs/Documenter.jl/blob/6c8679bac445ab1c656744075b13a57835081cc3/src/DocTests.jl#L339) should fix this issue, but there ought to be a better way.
-->

<details>

<summary> Here is the stacktrace of where select_pivot gets called by Documenter when running doctests: </summary>

```
  [2] select_pivot(lo::Int64, hi::Int64)
    @ Main.__doctest__named__Die ./none:3
  [3] partition!(t::Vector{Symbol}, lo::Int64, hi::Int64, o::Base.Order.ForwardOrdering, v::Vector{Symbol}, rev::Bool)
    @ Base.Sort ./sort.jl:544
  [4] sort!(v::Vector{Symbol}, lo::Int64, hi::Int64, a::PartialQuickSort{Missing, Missing}, o::Base.Order.ForwardOrdering, t::Vector{Symbol}, swap::Bool, rev::Bool)
    @ Base.Sort ./sort.jl:575
  [5] sort!
    @ ./sort.jl:574 [inlined]
  [6] sort!
    @ ./sort.jl:747 [inlined]
  [7] sort!
    @ ./sort.jl:859 [inlined]
  [8] #sort!#8
    @ ./sort.jl:914 [inlined]
  [9] sort!
    @ ./sort.jl:907 [inlined]
 [10] #names#8
    @ ./reflection.jl:102 [inlined]
 [11] names
    @ ./reflection.jl:102 [inlined]
 [12] make_typealias(x::Type)
    @ Base ./show.jl:600
 [13] show_typealias(io::IOContext{IOBuffer}, x::Type)
    @ Base ./show.jl:781
 [14] _show_type(io::IOContext{IOBuffer}, x::Type)
    @ Base ./show.jl:946
 [15] show(io::IOContext{IOBuffer}, x::Type)
    @ Base ./show.jl:941
 [16] print(io::IOContext{IOBuffer}, x::Type)
    @ Base ./strings/io.jl:35
 [17] showarg
    @ ./show.jl:2953 [inlined]
 [18] array_summary
    @ ./show.jl:2896 [inlined]
 [19] summary
    @ ./show.jl:2893 [inlined]
 [20] show(io::IOContext{IOBuffer}, #unused#::MIME{Symbol("text/plain")}, X::Vector{Main.__doctest__named__Die.Die})
    @ Base ./arrayshow.jl:368
 [21] #invokelatest#2
    @ ./essentials.jl:801 [inlined]
 [22] invokelatest
    @ ./essentials.jl:798 [inlined]
 [23] result_to_string
    @ ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:338 [inlined]
 [24] checkresult(sandbox::Module, result::Documenter.DocTests.Result, meta::Dict{Symbol, Any}, doc::Documenter.Documents.Document)
    @ Documenter.DocTests ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:318
 [25] eval_repl(block::Documenter.DocTests.MutableMD2CodeBlock, sandbox::Module, meta::Dict{Symbol, Any}, doc::Documenter.Documents.Document, page::String)
    @ Documenter.DocTests ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:247
 [26] doctest(ctx::Documenter.DocTests.DocTestContext, block_immutable::Documenter.Utilities.Markdown2.CodeBlock)
    @ Documenter.DocTests ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:185
 [27] (::Documenter.DocTests.var"#1#2"{Documenter.DocTests.DocTestContext})(node::Documenter.Utilities.Markdown2.CodeBlock)
    @ Documenter.DocTests ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:111
 [28] walk(f::Documenter.DocTests.var"#1#2"{Documenter.DocTests.DocTestContext}, node::Documenter.Utilities.Markdown2.CodeBlock)
    @ Documenter.Utilities.Markdown2 ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Utilities/Markdown2.jl:297
 [29] walk(f::Function, nodes::Vector{Documenter.Utilities.Markdown2.MarkdownBlockNode})
    @ Documenter.Utilities.Markdown2 ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Utilities/Markdown2.jl:306
 [30] walk
    @ ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Utilities/Markdown2.jl:299 [inlined]
 [31] doctest
    @ ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:108 [inlined]
 [32] doctest
    @ ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:66 [inlined]
 [33] doctest(blueprint::Documenter.Documents.DocumentBlueprint, doc::Documenter.Documents.Document)
    @ Documenter.DocTests ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/DocTests.jl:50
 [34] runner(#unused#::Type{Documenter.Builder.Doctest}, doc::Documenter.Documents.Document)
    @ Documenter.Builder ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Builder.jl:213
 [35] dispatch(#unused#::Type{Documenter.Builder.DocumentPipeline}, x::Documenter.Documents.Document)
    @ Documenter.Utilities.Selectors ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Utilities/Selectors.jl:170
 [36] #2
    @ ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Documenter.jl:267 [inlined]
 [37] cd(f::Documenter.var"#2#3"{Documenter.Documents.Document}, dir::String)
    @ Base.Filesystem ./file.jl:112
 [38] makedocs(; debug::Bool, format::Documenter.Writers.HTMLWriter.HTML, kwargs::Base.Pairs{Symbol, Any, NTuple{11, Symbol}, NamedTuple{(:build, :modules, :clean, :doctest, :linkcheck, :linkcheck_ignore, :strict, :checkdocs, :sitename, :authors, :pages), Tuple{String, Vector{Module}, Bool, Bool, Bool, Vector{String}, Bool, Symbol, String, String, Vector{Any}}}})
    @ Documenter ~/.julia/dev/julia/doc/deps/packages/Documenter/tBpqI/src/Documenter.jl:266
 [39] top-level scope
    @ ~/.julia/dev/julia/doc/make.jl:296
in expression starting at /Users/x/.julia/dev/julia/doc/make.jl:296
```

</details>

Reply:
How would you accommodate this in Documenter? Do you mean something like storing the RNG state, running `show`, and then resetting the state, as if `show` had not been called?

As an alternative, would using [doctest regex filters](https://juliadocs.github.io/Documenter.jl/dev/man/doctests/#Filtering-Doctests) be an option to work around this issue?

Reply:
> storing the RNG state, running `show`, and then resetting the state

That's the only way I can think of to do it. It's not perfect because if someone copied a show + RNG example from a REPL to a doctest it might behave differently, albeit consistently differently, as a doctest. Empirically, that edge case only comes up twice in all of JuliaLang/julia's CI. In those uncommon cases, regex filters or `doctest(fix=true)` should work as workarounds.

Reply:
x-ref https://github.com/JuliaDocs/Documenter.jl/pull/1869

Reply:
I like to think of sorting as pretty cheap, but yes, it can't hurt!